gkjohnson/three-mesh-bvh
# v0.4.0 changes

**TODO**
- [ ] Fix LGTM exclude
- [x] Update point cloud demo shapecast call, remove lazy generation option
- [x] Re-add rollup esm, umd to `/builds` folder for users that copy files
- [x] Add dev-bundle folder
- [x] Clean up sculpt example
- [x] Include an offset inclusive of the other geometry group nodes in the returned node indices for the sake of refit.
- [x] Sculpt: Refit performance improvement
- [x] Sculpt: Update triangle iteration function to take triangle index
- [x] Sculpt: Remove BVH helper from scene when not visible
- [x] Sculpt: Only update helper when not visible
- [x] Change shapecast to take object of callbacks (Fix #218)
- [x] Change shapecast to provide a triangle range callback (Fix #173)
- [x] `refit` tests
- [x] Add touch to collect triangles, sculpt examples
- [x] Retain index, position attribute references #206
    - remove geometry argument from "refit"
- [x] Only generate normals once at the end of all the steps
- [x] Consider computing the average normal and point in the triangle callback instead of after
- [x] Max steps to 10, maximum value to 25
- [x] Disable frustum culling and / or clear the bounding boxes for the target mesh when refitting the bounds
- [x] For flatten brush position the plane at the average point so we can include steps without refit
- [x] Flatten / smooth stroke
- [x] Sculpt comments
- [x] Check mouse distance movement in world space rather than screen space
- [x] Improve initial brush settings
- [x] Fix #186
- [x] Fix brush behavior
    - Use average points for the sculpt normal (average normals)
    - Iterate multiple steps with a threshold scaled by brush radius
- [x] Docs (~refit function~, ~sculpt example~) 
- [x] Make issue for sculpt example performance improvement (#207)
    - Give hints to refit function
    - Improve normal generation performance
    - Improve geometry generation performance
    - Add support for simplification and tessellation
- [x] Remove "buffer" from cast function names
- [x] Add `refitBounds` function (add benchmark)
- [x] Find clay matcap
- [ ] ~Update sculpt example to use range function~ (#229)
- [ ] ~root node does not fully contain child nodes after rebuild~ #222
    - it's possibly just the visualization helper?
    - Fixes itself on refit call
    - This is because the geometry bounding box is out of date and yields an incorrect root bounds. We should be able to guarantee a correct bounds by taking the union of the children or just removing the check for an existing bounds. Or update docs stating the bounding box is expected to be up to date?
- [ ] ~Perf could be improved by removing redundant vector reads. We only need unique triangle vertex indices out of the shapecast run to process later. #218~
    - Provide shapecast callback that just gets fired with triangle ranges
    - Adjust shapecast to take an object of functions so it's more convenient to call and log deprecation warning if the first argument is a function.
